src/mbgl/text/placement.cpp

```diff
@@ -0,0 +1,307 @@
+{}
+
+void Placement::placeLayer(RenderSymbolLayer& symbolLayer, const mat4& projMatrix, bool showCollisionBoxes) {
+
+    std::unordered_set<uint32_t> seenCrossTileIDs;
+
+    for (RenderTile& renderTile : symbolLayer.renderTiles) {
+
+        if (!renderTile.tile.isRenderable()) {
+            continue;
+        }
+
+        auto bucket = renderTile.tile.getBucket(*symbolLayer.baseImpl);
+        assert(dynamic_cast<SymbolBucket*>(bucket));
+        SymbolBucket& symbolBucket = *reinterpret_cast<SymbolBucket*>(bucket);
+
+        auto& layout = symbolBucket.layout;
+
+        const float pixelsToTileUnits = renderTile.id.pixelsToTileUnits(1, state.getZoom());
+
+        const float scale = std::pow(2, state.getZoom() - renderTile.tile.id.overscaledZ);
+        const float textPixelRatio = util::EXTENT / (util::tileSize * renderTile.tile.id.overscaleFactor());
+
+        mat4 posMatrix;
+        state.matrixFor(posMatrix, renderTile.id);
+        matrix::multiply(posMatrix, projMatrix, posMatrix);
+
+        mat4 textLabelPlaneMatrix = getLabelPlaneMatrix(posMatrix,
+                layout.get<TextPitchAlignment>() == style::AlignmentType::Map,
+                layout.get<TextRotationAlignment>() == style::AlignmentType::Map,
+                state,
+                pixelsToTileUnits);
+
+        mat4 iconLabelPlaneMatrix = getLabelPlaneMatrix(posMatrix,
+                layout.get<IconPitchAlignment>() == style::AlignmentType::Map,
+                layout.get<IconRotationAlignment>() == style::AlignmentType::Map,
+                state,
+                pixelsToTileUnits);
+
+        placeLayerBucket(symbolBucket, posMatrix, textLabelPlaneMatrix, iconLabelPlaneMatrix, scale, textPixelRatio, showCollisionBoxes, seenCrossTileIDs);
+    }
+}
+
+void Placement::placeLayerBucket(
+        SymbolBucket& bucket,
+        const mat4& posMatrix,
+        const mat4& textLabelPlaneMatrix,
+        const mat4& iconLabelPlaneMatrix,
+        const float scale,
+        const float textPixelRatio,
+        const bool showCollisionBoxes,
+        std::unordered_set<uint32_t>& seenCrossTileIDs) {
+
+    auto partiallyEvaluatedTextSize = bucket.textSizeBinder->evaluateForZoom(state.getZoom());
+    auto partiallyEvaluatedIconSize = bucket.iconSizeBinder->evaluateForZoom(state.getZoom());
+
+    const bool iconWithoutText = !bucket.hasTextData() || bucket.layout.get<TextOptional>();
+    const bool textWithoutIcon = !bucket.hasIconData() || bucket.layout.get<IconOptional>();
+
+    for (auto& symbolInstance : bucket.symbolInstances) {
+
+        if (seenCrossTileIDs.count(symbolInstance.crossTileID) == 0) {
+            bool placeText = false;
+            bool placeIcon = false;
+
+            if (symbolInstance.placedTextIndex) {
+                PlacedSymbol& placedSymbol = bucket.text.placedSymbols.at(*symbolInstance.placedTextIndex);
+                const float fontSize = evaluateSizeForFeature(partiallyEvaluatedTextSize, placedSymbol);
```

Comment:
We don't use the font size for point features, so we could avoid the calculation here. The fact that we don't dynamically evaluated font size for point feature collision boxes means they're a little off from their ideal size (and if a label has its font size changing with zoom, you can see a little jump in the box as you cross integer zoom levels), but the difference is usually pretty minor.

Reply:
> The fact that we don't dynamically evaluated font size for point feature collision boxes means they're a little off from their ideal size

Should we use the exact size?

Reply:
🤷‍♂️ I'm not sure. It seems like a minor fidelity win in exchange for a minor performance cost.